# Bag of words

In [42]:
import pandas as pd
d = pd.read_csv('https://raw.githubusercontent.com/hjyoon02/hcmld21/main/Datasets/Youtube01-Psy.csv')

In [43]:
d.tail()

COMMENT_ID  ... CLASS
345      z13th1q4yzihf1bll23qxzpjeujterydj  ...     0
346  z13fcn1wfpb5e51xe04chdxakpzgchyaxzo0k  ...     0
347    z130zd5b3titudkoe04ccbeohojxuzppvbg  ...     1
348    z12he50arvrkivl5u04cctawgxzkjfsjcc4  ...     1
349  z13vhvu54u3ewpp5h04ccb4zuoardrmjlyk0k  ...     0

[5 rows x 5 columns]

In [44]:
len(d.query('CLASS == 1'))

175

In [45]:
len(d.query('CLASS == 0'))

175

In [46]:
len(d)

350

In [47]:
from sklearn.feature_extraction.text import CountVectorizer
vectorizer = CountVectorizer()

In [48]:
dvec = vectorizer.fit_transform(d['CONTENT'])

In [49]:
dvec

<350x1418 sparse matrix of type '<class 'numpy.int64'>'
	with 4354 stored elements in Compressed Sparse Row format>

In [50]:
print(d['CONTENT'][349])
analyze = vectorizer.build_analyzer()
analyze(d['CONTENT'][349])

The first billion viewed this because they thought it was really cool, the  other billion and a half came to see how stupid the first billion were...


['the',
 'first',
 'billion',
 'viewed',
 'this',
 'because',
 'they',
 'thought',
 'it',
 'was',
 'really',
 'cool',
 'the',
 'other',
 'billion',
 'and',
 'half',
 'came',
 'to',
 'see',
 'how',
 'stupid',
 'the',
 'first',
 'billion',
 'were']

In [51]:
vectorizer.get_feature_names()

['00',
 '000',
 '02',
 '034',
 '05',
 '08',
 '10',
 '100',
 '100000415527985',
 '10200253113705769',
 '1030',
 '1073741828',
 '11',
 '1111',
 '112720997191206369631',
 '12',
 '123',
 '124',
 '124923004',
 '126',
 '127',
 '13017194',
 '131338190916',
 '1340488',
 '1340489',
 '1340490',
 '1340491',
 '1340492',
 '1340493',
 '1340494',
 '1340499',
 '1340500',
 '1340502',
 '1340503',
 '1340504',
 '1340517',
 '1340518',
 '1340519',
 '1340520',
 '1340521',
 '1340522',
 '1340523',
 '1340524',
 '134470083389909',
 '1415297812',
 '1495323920744243',
 '1496241863981208',
 '1496273723978022',
 '1498561870415874',
 '161620527267482',
 '171183229277',
 '19',
 '19924',
 '1firo',
 '1m',
 '20',
 '2009',
 '2012',
 '2012bitches',
 '2013',
 '2014',
 '201470069872822',
 '2015',
 '2017',
 '210',
 '23',
 '24',
 '24398',
 '243a',
 '279',
 '29',
 '2b',
 '2billion',
 '2x10',
 '300',
 '3000',
 '313327',
 '315',
 '322',
 '33',
 '33gxrf',
 '39',
 '390875584405933',
 '391725794320912',
 '40beuutvu2zkxk4utgpz8k',
 '

In [52]:
dshuf = d.sample(frac=1)

In [53]:
d_train = dshuf[:300]
d_test = dshuf[300:]
d_train_att = vectorizer.fit_transform(d_train['CONTENT']) # fit bag-of-words on training set
d_test_att = vectorizer.transform(d_test['CONTENT']) # reuse on testing set
d_train_label = d_train['CLASS']
d_test_label = d_test['CLASS']

In [54]:
d_train_att

<300x1259 sparse matrix of type '<class 'numpy.int64'>'
	with 3717 stored elements in Compressed Sparse Row format>

In [55]:
d_test_att

<50x1259 sparse matrix of type '<class 'numpy.int64'>'
	with 474 stored elements in Compressed Sparse Row format>

In [56]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier(n_estimators=80)

In [57]:
clf.fit(d_train_att, d_train_label)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=80,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

In [58]:
clf.score(d_test_att, d_test_label)

0.96

In [59]:
from sklearn.metrics import confusion_matrix
pred_labels = clf.predict(d_test_att)
confusion_matrix(d_test_label, pred_labels)

array([[27,  0],
       [ 2, 21]])

In [60]:
from sklearn.model_selection import cross_val_score
scores = cross_val_score(clf, d_train_att, d_train_label, cv=5)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.95 (+/- 0.04)


In [61]:
d = pd.concat([pd.read_csv('https://raw.githubusercontent.com/hjyoon02/hcmld21/main/Datasets/Youtube01-Psy.csv'),
               pd.read_csv('https://raw.githubusercontent.com/hjyoon02/hcmld21/main/Datasets/Youtube02-KatyPerry.csv'),
               pd.read_csv('https://raw.githubusercontent.com/hjyoon02/hcmld21/main/Datasets/Youtube03-LMFAO.csv'),
               pd.read_csv('https://raw.githubusercontent.com/hjyoon02/hcmld21/main/Datasets/Youtube04-Eminem.csv'),
               pd.read_csv('https://raw.githubusercontent.com/hjyoon02/hcmld21/main/Datasets/Youtube05-Shakira.csv')])

In [62]:
len(d)

1956

In [63]:
len(d.query('CLASS == 1'))

1005

In [64]:
len(d.query('CLASS == 0'))

951

In [65]:
dshuf = d.sample(frac=1)
d_content = dshuf['CONTENT']
d_label = dshuf['CLASS']

In [66]:
# set up a pipeline
from sklearn.pipeline import Pipeline, make_pipeline
pipeline = Pipeline([
                     ('bag-of-words', CountVectorizer()),
                     ('random forest', RandomForestClassifier()),
])
pipeline

Pipeline(memory=None,
         steps=[('bag-of-words',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, vocabu...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

In [67]:
pipeline = make_pipeline(CountVectorizer(), RandomForestClassifier())

In [68]:
pipeline.fit(d_content[:1500], d_label[:1500])

Pipeline(memory=None,
         steps=[('countvectorizer',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=None, min_df=1,
                                 ngram_range=(1, 1), preprocessor=None,
                                 stop_words=None, strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=None, voc...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                

In [69]:
pipeline.score(d_content[1500:], d_label[1500:])

0.9605263157894737

In [70]:
pipeline.predict(["what a neat video!"])

array([0])

In [71]:
pipeline.predict(["plz subscribe to my channel"])

array([1])

In [72]:
scores = cross_val_score(pipeline, d_content, d_label, cv=5)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.96 (+/- 0.01)


In [73]:
# add TF-IDF
from sklearn.feature_extraction.text import TfidfTransformer
pipeline2 = make_pipeline(CountVectorizer(),
                          TfidfTransformer(norm=None),
                          RandomForestClassifier())

In [74]:
scores = cross_val_score(pipeline2, d_content, d_label, cv=5)
print("Accuracy: %0.2f (+/- %0.2f)" % (scores.mean(), scores.std() * 2))

Accuracy: 0.96 (+/- 0.01)


In [75]:
pipeline2.steps

[('countvectorizer',
  CountVectorizer(analyzer='word', binary=False, decode_error='strict',
                  dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                  lowercase=True, max_df=1.0, max_features=None, min_df=1,
                  ngram_range=(1, 1), preprocessor=None, stop_words=None,
                  strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                  tokenizer=None, vocabulary=None)),
 ('tfidftransformer',
  TfidfTransformer(norm=None, smooth_idf=True, sublinear_tf=False, use_idf=True)),
 ('randomforestclassifier',
  RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                         criterion='gini', max_depth=None, max_features='auto',
                         max_leaf_nodes=None, max_samples=None,
                         min_impurity_decrease=0.0, min_impurity_split=None,
                         min_samples_leaf=1, min_samples_split=2,
                         min_weight_fraction_leaf=0.0, 

In [76]:
# parameter search
parameters = {
    'countvectorizer__max_features': (None, 1000, 2000),
    'countvectorizer__ngram_range': ((1, 1), (1, 2)),  # unigrams or bigrams
    'countvectorizer__stop_words': ('english', None),
    'tfidftransformer__use_idf': (True, False),  # effectively turn on/off tfidf
    'randomforestclassifier__n_estimators': (20, 50, 100)
}
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(pipeline2, parameters, n_jobs=-1, verbose=1)

In [77]:
grid_search.fit(d_content, d_label)

Fitting 5 folds for each of 72 candidates, totalling 360 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:   13.3s
[Parallel(n_jobs=-1)]: Done 196 tasks      | elapsed:  1.0min
[Parallel(n_jobs=-1)]: Done 360 out of 360 | elapsed:  1.8min finished


GridSearchCV(cv=None, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('countvectorizer',
                                        CountVectorizer(analyzer='word',
                                                        binary=False,
                                                        decode_error='strict',
                                                        dtype=<class 'numpy.int64'>,
                                                        encoding='utf-8',
                                                        input='content',
                                                        lowercase=True,
                                                        max_df=1.0,
                                                        max_features=None,
                                                        min_df=1,
                                                        ngram_range=(1, 1),
                                              

In [78]:
print("Best score: %0.3f" % grid_search.best_score_)
print("Best parameters set:")
best_parameters = grid_search.best_estimator_.get_params()
for param_name in sorted(parameters.keys()):
  print("\t%s: %r" % (param_name, best_parameters[param_name]))

Best score: 0.962
Best parameters set:
	countvectorizer__max_features: 2000
	countvectorizer__ngram_range: (1, 2)
	countvectorizer__stop_words: 'english'
	randomforestclassifier__n_estimators: 100
	tfidftransformer__use_idf: True


# 4.2
I visited this site: https://www.youtube.com/watch?v=v7bnOxV4jAc

answer : 0, 1, 0, 1, 0

predict: 0, 0, 1, 1, 0

For my 5 test cases, the model's accuracy is 60%.
It may be because of tricky comments.

The second one is a spam comment, but actually it had been a normal comment until she add a sentence "ONLYFANS MEETME.BABY." (caution: You shouldn't visit this site!) So the model may predict it is not a spam comment.

The third one is not a spam comment, but the model predicts it is a spam comment. If this video wasn't IU's movie, I would classify it as a spam comment. No one knows it is a spam comment or not without knowing what the vedio is. But our data set doen't include the information of videos. Therefore, it is not awkward for the prediction wrong.

In [83]:
pipeline.predict(["BBIBBI (2018) - YELLOW BLUEMING (2019) - BLUE EIGHT (2020) - ORANGE LILAC (2021) - LILAC"])

array([0])

In [84]:
pipeline.predict(['ONLYFANS MEETME.BABY Youtube: This is fine Someone: Says "heck" Youtube: BE GONE'])

array([0])

In [85]:
pipeline.predict(["If I had IU's voice then, I would never shut my mouth"])

array([1])

In [86]:
pipeline.predict(["If you are a fan of IU then watch my new video. I made a beautiful cartoon picture of IU"])

array([1])

In [87]:
pipeline.predict(["SHE'S FEEDING US SO WELL!!!"])

array([0])

# 4.3

In [96]:
# parameter search
parameters = {
    'countvectorizer__max_features': (None, 1000, 2000),
    'countvectorizer__ngram_range': ((1, 1), (1, 2), (1, 3)),  # unigram or bigram or trigram
    'countvectorizer__stop_words': ('english', None),
    'tfidftransformer__use_idf': (True, False),  # effectively turn on/off tfidf
    'randomforestclassifier__n_estimators': (20, 50, 100)
}

grid_search = GridSearchCV(pipeline2, parameters, n_jobs=-1, verbose=1)

In [97]:
grid_search.fit(d_content, d_label)

Fitting 5 folds for each of 108 candidates, totalling 540 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=-1)]: Done  46 tasks      | elapsed:   11.5s
[Parallel(n_jobs=-1)]: Done 196 tasks      | elapsed:  1.4min
[Parallel(n_jobs=-1)]: Done 446 tasks      | elapsed:  2.5min
[Parallel(n_jobs=-1)]: Done 540 out of 540 | elapsed:  3.1min finished


GridSearchCV(cv=None, error_score=nan,
             estimator=Pipeline(memory=None,
                                steps=[('countvectorizer',
                                        CountVectorizer(analyzer='word',
                                                        binary=False,
                                                        decode_error='strict',
                                                        dtype=<class 'numpy.int64'>,
                                                        encoding='utf-8',
                                                        input='content',
                                                        lowercase=True,
                                                        max_df=1.0,
                                                        max_features=None,
                                                        min_df=1,
                                                        ngram_range=(1, 1),
                                              

# 4.4
The best parameters set is:

	countvectorizer__max_features: 2000
	countvectorizer__ngram_range: (1, 2)
	countvectorizer__stop_words: 'english'
	randomforestclassifier__n_estimators: 100
	tfidftransformer__use_idf: True

So I can apply the result like the below code.

The score of 'pipeline' is 0.9605263157894737. Whereas, the score of 'pipeline3' is 0.9627192982456141, which is a little bit higher than 'pipeline'.

In [108]:
pipeline3 = make_pipeline(CountVectorizer(
    max_features=2000, ngram_range=(1, 2), stop_words='english'),
                          TfidfTransformer(use_idf=True), # turn on tfidf
                          RandomForestClassifier(n_estimators=100))

In [111]:
pipeline3.fit(d_content[:1500], d_label[:1500])

Pipeline(memory=None,
         steps=[('countvectorizer',
                 CountVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.int64'>, encoding='utf-8',
                                 input='content', lowercase=True, max_df=1.0,
                                 max_features=2000, min_df=1,
                                 ngram_range=(1, 2), preprocessor=None,
                                 stop_words='english', strip_accents=None,
                                 token_pattern='(?u)\\b\\w\\w+\\b',
                                 tokenizer=Non...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None,
                                 

In [112]:
pipeline3.score(d_content[1500:], d_label[1500:])

0.9627192982456141

# Word2Vec models

In [1]:
import gensim, logging

In [41]:
logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)

In [ ]:
gmodel = gensim.models.KeyedVectors.load_word2vec_format('GoogleNews-vectors-negative300.bin', binary=True)
# I cannot find the file